### ISSUE TITLE
httptrace support

### ISSUE THREAD

**Issue by daog1** [quic-go/quic-go#3342]
Recently, I was using http3 for server development, and then I tested it with K6. I found that K6 does not support http3, Then I joined guic to K6. Quic can't support httptrace, and then I can't do a very detailed performance test. Please support it
PR here:

https://github.com/grafana/k6/pull/2407

**Comment by marten-seemann** [comment]
How widely is httptrace actually used? Just wondering if it's worth the added complexity. What's the data that can be learned from this?

I'm not opposed to implementing this, although we won't have the bandwidth to do that ourselves. I might be convinced to review a PR though.

**Comment by daodao2007** [comment]
Here is some information about httptrace，
https://go.dev/blog/http-tracing
Go supports this feature in version 1.7
There are many benefits. HTTP and http2 have already realized this

**Comment by marten-seemann** [comment]
I saw that blog post, but one blog post from more than half a decade ago seems like a rather slim motivation for this.

Again, if there’s a strong use case for this, I wouldn’t be opposed to reviewing a PR, but don’t expect any effort being put into this from our side.

**Comment by iljavs** [comment]
FYI for those who need httptrace support -which I needed- you can manually perform the httptrace callouts by setting your own Dial (and calling DialAddrEarlyContext from there). Most info you need is provided. I did have to add the http request to the context so I could access the headers and provide them to WroteHeaderField.

**Comment by costela** [comment]
Can we please keep this issue open? I understand not having the bandwidth to implement this, but it's still a valuable addition. Case in point: 
- [prometheus uses httptrace](https://github.com/prometheus/client_golang/blob/main/prometheus/promhttp/instrument_client.go#L163) to provide client metrics.
- [opentelemetry uses httptrace] to enrich traces with client metrics.

Both are valuable to support observability of services using QUIC.

Not to mention k6, as OP pointed out, which is also a pretty widely adopted tool and which would considerably profit from QUIC support.

**Comment by marten-seemann** [comment]
Makes sense. Reopening.
I won't have the bandwidth to work on this any time soon, but I'd be happy to review a PR. Please comment on this issue first to discuss the design.

**Comment by 9bany** [comment]
Can i working on this @marten-seemann ?

**Comment by marten-seemann** [comment]
@9bany Sure. What will be your implementation strategy? Which tracer calls would you implement first?

**Comment by 9bany** [comment]
Yeah, I think I need to implement which tracer calls first. Since I don't have deep knowledge in `quic-go`, I need your help. Do you have any ideas on how to implement this feature? @marten-seemann

**Comment by letian0805** [comment]
Hi！Is there any progress on this?

**Comment by marten-seemann** [comment]
@lRoccoon put in the work and implemented this in #4749. Everyone who requested this feature, it would be highly appreciated if you could try out this PR and give feedback there! Thank you!
